File: ts-sdk/src/sdk/sdk.ts
import * as utils from "../internal/utils";
import { DatabaseBranches } from "./databasebranches";
import { DatabaseBranchPasswords } from "./databasebranchpasswords";
import { Databases } from "./databases";
import { DeployRequests } from "./deployrequests";
import { Security } from "./models/shared";
import { OAuthApplications } from "./oauthapplications";
import { OAuthTokens } from "./oauthtokens";
import { Organizations } from "./organizations";
import { Users } from "./users";
import axios, { AxiosInstance } from "axios";

export const ServerList = [
	"https://api.planetscale.com/v1",
] as const;



export type SDKProps = {
  defaultClient?: AxiosInstance;

  security?: Security;

  serverUrl?: string;
}


export class PlanetscaleAPI {
  public databaseBranchPasswords: DatabaseBranchPasswords;
  public databaseBranches: DatabaseBranches;
  public databases: Databases;
  public deployRequests: DeployRequests;
  public oAuthApplications: OAuthApplications;
  public oAuthTokens: OAuthTokens;
  public organizations: Organizations;
  public users: Users;

  public _defaultClient: AxiosInstance;
  public _securityClient: AxiosInstance;
  public _serverURL: string;
  private _language = "typescript";
  private _sdkVersion = "2.1.1";
  private _genVersion = "1.1.2";

  constructor(props: SDKProps) {
    this._serverURL = props.serverUrl ?? ServerList[0];

    this._defaultClient = props.defaultClient ?? axios.create({ baseURL: this._serverURL });
    if (props.security) {
      let security: Security = props.security;
      if (!(props.security instanceof utils.SpeakeasyBase))
        security = new Security(props.security);
      this._securityClient = utils.createSecurityClient(
        this._defaultClient,
        security
      );
    } else {
      this._securityClient = this._defaultClient;
    }
    
    this.databaseBranchPasswords = new DatabaseBranchPasswords(
      this._defaultClient,
      this._securityClient,
      this._serverURL,
      this._language,
      this._sdkVersion,
      this._genVersion
    );
    
    this.databaseBranches = new DatabaseBranches(
      this._defaultClient,
      this._securityClient,
      this._serverURL,
      this._language,
      this._sdkVersion,
      this._genVersion
    );
    
    this.databases = new Databases(
      this._defaultClient,
      this._securityClient,
      this._serverURL,
      this._language,
      this._sdkVersion,
      this._genVersion
    );
    
    this.deployRequests = new DeployRequests(
      this._defaultClient,
      this._securityClient,
      this._serverURL,
      this._language,
      this._sdkVersion,
      this._genVersion
    );
    
    this.oAuthApplications = new OAuthApplications(
      this._defaultClient,
      this._securityClient,
      this._serverURL,
      this._language,
      this._sdkVersion,
      this._genVersion
    );
    
    this.oAuthTokens = new OAuthTokens(
      this._defaultClient,
      this._securityClient,
      this._serverURL,
      this._language,
      this._sdkVersion,
      this._genVersion
    );
    
    this.organizations = new Organizations(
      this._defaultClient,
      this._securityClient,
      this._serverURL,
      this._language,
      this._sdkVersion,
      this._genVersion
    );
    
    this.users = new Users(
      this._defaultClient,
      this._securityClient,
      this._serverURL,
      this._language,
      this._sdkVersion,
      this._genVersion
    );
  }
  
}

File: go-sdk/sdk.go
package planetscale

import (
	"net/http"

	"github.com/speakeasy-sdks/planetscale-sdks/go-client-sdk/pkg/models/shared"
	"github.com/speakeasy-sdks/planetscale-sdks/go-client-sdk/pkg/utils"
)

var ServerList = []string{
	"https://api.planetscale.com/v1",
}

type HTTPClient interface {
	Do(req *http.Request) (*http.Response, error)
}

type Planetscale struct {
	DatabaseBranchPasswords *DatabaseBranchPasswords
	DatabaseBranches        *DatabaseBranches
	Databases               *Databases
	DeployRequests          *DeployRequests
	OAuthApplications       *OAuthApplications
	OAuthTokens             *OAuthTokens
	Organizations           *Organizations
	Users                   *Users

	_defaultClient  HTTPClient
	_securityClient HTTPClient
	_security       *shared.Security
	_serverURL      string
	_language       string
	_sdkVersion     string
	_genVersion     string
}

type SDKOption func(*Planetscale)

func WithServerURL(serverURL string, params map[string]string) SDKOption {
	return func(sdk *Planetscale) {
		if params != nil {
			serverURL = utils.ReplaceParameters(serverURL, params)
		}

		sdk._serverURL = serverURL
	}
}

func WithClient(client HTTPClient) SDKOption {
	return func(sdk *Planetscale) {
		sdk._defaultClient = client
	}
}

func WithSecurity(security shared.Security) SDKOption {
	return func(sdk *Planetscale) {
		sdk._security = &security
	}
}

func New(opts ...SDKOption) *Planetscale {
	sdk := &Planetscale{
		_language:   "go",
		_sdkVersion: "1.2.0",
		_genVersion: "0.22.0",
	}
	for _, opt := range opts {
		opt(sdk)
	}

	if sdk._defaultClient == nil {
		sdk._defaultClient = http.DefaultClient
	}
	if sdk._securityClient == nil {

		if sdk._security != nil {
			sdk._securityClient = utils.ConfigureSecurityClient(sdk._defaultClient, sdk._security)
		} else {
			sdk._securityClient = sdk._defaultClient
		}

	}

	if sdk._serverURL == "" {
		sdk._serverURL = ServerList[0]
	}

	sdk.DatabaseBranchPasswords = NewDatabaseBranchPasswords(
		sdk._defaultClient,
		sdk._securityClient,
		sdk._serverURL,
		sdk._language,
		sdk._sdkVersion,
		sdk._genVersion,
	)

	sdk.DatabaseBranches = NewDatabaseBranches(
		sdk._defaultClient,
		sdk._securityClient,
		sdk._serverURL,
		sdk._language,
		sdk._sdkVersion,
		sdk._genVersion,
	)

	sdk.Databases = NewDatabases(
		sdk._defaultClient,
		sdk._securityClient,
		sdk._serverURL,
		sdk._language,
		sdk._sdkVersion,
		sdk._genVersion,
	)

	sdk.DeployRequests = NewDeployRequests(
		sdk._defaultClient,
		sdk._securityClient,
		sdk._serverURL,
		sdk._language,
		sdk._sdkVersion,
		sdk._genVersion,
	)

	sdk.OAuthApplications = NewOAuthApplications(
		sdk._defaultClient,
		sdk._securityClient,
		sdk._serverURL,
		sdk._language,
		sdk._sdkVersion,
		sdk._genVersion,
	)

	sdk.OAuthTokens = NewOAuthTokens(
		sdk._defaultClient,
		sdk._securityClient,
		sdk._serverURL,
		sdk._language,
		sdk._sdkVersion,
		sdk._genVersion,
	)

	sdk.Organizations = NewOrganizations(
		sdk._defaultClient,
		sdk._securityClient,
		sdk._serverURL,
		sdk._language,
		sdk._sdkVersion,
		sdk._genVersion,
	)

	sdk.Users = NewUsers(
		sdk._defaultClient,
		sdk._securityClient,
		sdk._serverURL,
		sdk._language,
		sdk._sdkVersion,
		sdk._genVersion,
	)

	return sdk
}
